File: app/control_plane/document_ingestion.py
from __future__ import annotations

import hashlib
import os
import re
import shutil
import subprocess
import tempfile
import zipfile
from dataclasses import dataclass
from io import BytesIO
from pathlib import Path

from docx import Document


MAX_DOCUMENTS = 3
MAX_DOCUMENT_BYTES = 5 * 1024 * 1024
MAX_TOTAL_BYTES = 10 * 1024 * 1024
MAX_EXTRACTED_CHARACTERS = 120_000
MAX_DOCX_MEMBERS = 500
MAX_DOCX_UNCOMPRESSED_BYTES = 25 * 1024 * 1024
SUPPORTED_EXTENSIONS = (".doc", ".docx", ".txt")


class DocumentIngestionError(ValueError):
    pass


@dataclass(frozen=True, slots=True)
class DocumentSection:
    reference: str
    heading: str
    text: str


@dataclass(frozen=True, slots=True)
class ExtractedDocument:
    id: str
    name: str
    format: str
    size_bytes: int
    sha256: str
    character_count: int
    sections: tuple[DocumentSection, ...]

    def analysis_text(self) -> str:
        lines = [f"DOCUMENT {self.id}: {self.name}"]
        for section in self.sections:
            heading = f" | {section.heading}" if section.heading else ""
            lines.append(f"[SOURCE {section.reference}{heading}]")
            lines.append(section.text)
        return "\n".join(lines)

    def public_payload(self) -> dict[str, object]:
        return {
            "id": self.id,
            "name": self.name,
            "format": self.format,
            "size_bytes": self.size_bytes,
            "sha256": self.sha256,
            "character_count": self.character_count,
            "section_count": len(self.sections),
        }


def extract_documents(files: tuple[tuple[str, bytes], ...]) -> tuple[ExtractedDocument, ...]:
    if not files:
        raise DocumentIngestionError("Select at least one document to analyze.")
    if len(files) > MAX_DOCUMENTS:
        raise DocumentIngestionError(f"Upload no more than {MAX_DOCUMENTS} documents.")
    total_bytes = sum(len(content) for _, content in files)
    if total_bytes > MAX_TOTAL_BYTES:
        raise DocumentIngestionError("The combined document size exceeds 10 MB.")

    documents = tuple(
        extract_document(index=index, filename=filename, content=content)
        for index, (filename, content) in enumerate(files, start=1)
    )
    extracted_characters = sum(item.character_count for item in documents)
    if extracted_characters > MAX_EXTRACTED_CHARACTERS:
        raise DocumentIngestionError(
            "The documents contain too much text for one analysis. Split them into smaller files."
        )
    return documents


def extract_document(*, index: int, filename: str, content: bytes) -> ExtractedDocument:
    safe_name = Path(filename or "document").name.strip()
    extension = Path(safe_name).suffix.casefold()
    if extension not in SUPPORTED_EXTENSIONS:
        raise DocumentIngestionError(
            "Unsupported document type. Upload Word (.doc or .docx) or plain text (.txt)."
        )
    if not content:
        raise DocumentIngestionError(f"{safe_name} is empty.")
    if len(content) > MAX_DOCUMENT_BYTES:
        raise DocumentIngestionError(f"{safe_name} exceeds the 5 MB file limit.")

    document_id = f"document-{index}"
    if extension == ".docx":
        sections = _docx_sections(document_id, safe_name, content)
        format_name = "docx"
    elif extension == ".doc":
        sections = _legacy_doc_sections(document_id, safe_name, content)
        format_name = "doc"
    else:
        sections = _text_sections(document_id, safe_name, content)
        format_name = "txt"

    character_count = sum(len(item.text) for item in sections)
    if character_count < 20:
        raise DocumentIngestionError(
            f"{safe_name} does not contain enough readable text to analyze."
        )
    if character_count > MAX_EXTRACTED_CHARACTERS:
        raise DocumentIngestionError(
            f"{safe_name} contains too much text for one analysis. Split it into smaller files."
        )
    return ExtractedDocument(
        id=document_id,
        name=safe_name,
        format=format_name,
        size_bytes=len(content),
        sha256=hashlib.sha256(content).hexdigest(),
        character_count=character_count,
        sections=sections,
    )


def _docx_sections(
    document_id: str, filename: str, content: bytes
) -> tuple[DocumentSection, ...]:
    if not content.startswith(b"PK"):
        raise DocumentIngestionError(f"{filename} is not a valid DOCX document.")
    try:
        with zipfile.ZipFile(BytesIO(content)) as archive:
            members = archive.infolist()
            if len(members) > MAX_DOCX_MEMBERS:
                raise DocumentIngestionError(f"{filename} contains too many embedded files.")
            if any(item.flag_bits & 0x1 for item in members):
                raise DocumentIngestionError(f"Encrypted DOCX files are not supported: {filename}.")
            if sum(item.file_size for item in members) > MAX_DOCX_UNCOMPRESSED_BYTES:
                raise DocumentIngestionError(f"{filename} expands beyond the safe DOCX limit.")
        document = Document(BytesIO(content))
    except DocumentIngestionError:
        raise
    except (OSError, ValueError, zipfile.BadZipFile) as error:
        raise DocumentIngestionError(f"{filename} could not be read as DOCX.") from error

    sections: list[DocumentSection] = []
    current_heading = ""
    paragraph_index = 0
    for paragraph in document.paragraphs:
        text = _clean_text(paragraph.text)
        if not text:
            continue
        paragraph_index += 1
        style_name = (paragraph.style.name if paragraph.style else "").casefold()
        if style_name.startswith("heading") or style_name in {"title", "subtitle"}:
            current_heading = text
        sections.append(
            DocumentSection(
                reference=f"{document_id}:paragraph-{paragraph_index}",
                heading=current_heading,
                text=text,
            )
        )
    for table_index, table in enumerate(document.tables, start=1):
        for row_index, row in enumerate(table.rows, start=1):
            values = [_clean_text(cell.text) for cell in row.cells]
            text = " | ".join(value for value in values if value)
            if text:
                sections.append(
                    DocumentSection(
                        reference=f"{document_id}:table-{table_index}-row-{row_index}",
                        heading=current_heading,
                        text=text,
                    )
                )
    return tuple(sections)


def _legacy_doc_sections(
    document_id: str, filename: str, content: bytes
) -> tuple[DocumentSection, ...]:
    if not content.startswith(bytes.fromhex("D0CF11E0A1B11AE1")):
        raise DocumentIngestionError(f"{filename} is not a valid legacy Word document.")
    executable = shutil.which("antiword")
    converter = "antiword"
    if executable is None and os.uname().sysname == "Darwin":
        executable = shutil.which("textutil")
        converter = "textutil"
    if executable is None:
        raise DocumentIngestionError(
            "Legacy .doc conversion is unavailable on this server. Save the file as .docx and retry."
        )
    try:
        with tempfile.TemporaryDirectory(prefix="tasklattice-doc-") as directory:
            source = Path(directory) / "source.doc"
            source.write_bytes(content)
            command = (
                [executable, str(source)]
                if converter == "antiword"
                else [executable, "-convert", "txt", "-stdout", str(source)]
            )
            result = subprocess.run(
                command,
                cwd=directory,
                stdin=subprocess.DEVNULL,
                capture_output=True,
                check=False,
                timeout=10,
            )
    except (OSError, subprocess.SubprocessError) as error:
        raise DocumentIngestionError(f"{filename} could not be converted safely.") from error
    if result.returncode != 0 or not result.stdout.strip():
        raise DocumentIngestionError(f"{filename} could not be read as a Word document.")
    return _text_sections(document_id, filename, result.stdout)


def _text_sections(
    document_id: str, filename: str, content: bytes
) -> tuple[DocumentSection, ...]:
    text = _decode_text(filename, content)
    lines = text.splitlines()
    sections: list[DocumentSection] = []
    for start in range(0, len(lines), 20):
        chunk = _clean_text("\n".join(lines[start : start + 20]))
        if not chunk:
            continue
        end = min(start + 20, len(lines))
        sections.append(
            DocumentSection(
                reference=f"{document_id}:lines-{start + 1}-{end}",
                heading="",
                text=chunk,
            )
        )
    return tuple(sections)


def _decode_text(filename: str, content: bytes) -> str:
    encodings = ("utf-8-sig", "utf-16", "gb18030")
    for encoding in encodings:
        try:
            value = content.decode(encoding)
        except UnicodeDecodeError:
            continue
        if "\x00" not in value:
            return value
    raise DocumentIngestionError(f"{filename} is not valid UTF-8, UTF-16, or GB18030 text.")


def _clean_text(value: str) -> str:
    lines = [re.sub(r"[ \t]+", " ", line).strip() for line in value.splitlines()]
    return "\n".join(line for line in lines if line).strip()

File: controller/server/control-plane-ai/document-ingestion.ts
import { execFile } from "node:child_process";
import { createHash } from "node:crypto";
import { mkdtemp, rm, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { basename, extname, join } from "node:path";
import { promisify } from "node:util";

import JSZip from "jszip";
import mammoth from "mammoth";

import { ValidationError } from "../domain/errors.js";

export const MAX_DOCUMENTS = 3;
export const MAX_DOCUMENT_BYTES = 5 * 1024 * 1024;
export const MAX_TOTAL_BYTES = 10 * 1024 * 1024;
const MAX_EXTRACTED_CHARACTERS = 120_000;
const MAX_DOCX_MEMBERS = 500;
const MAX_DOCX_UNCOMPRESSED_BYTES = 25 * 1024 * 1024;

export type DocumentSection = { reference: string; heading: string; text: string };
export type ExtractedDocument = {
  id: string;
  name: string;
  format: "doc" | "docx" | "txt";
  size_bytes: number;
  sha256: string;
  character_count: number;
  section_count: number;
  sections: DocumentSection[];
};

const execute = promisify(execFile);

export async function extractDocuments(files: File[]): Promise<ExtractedDocument[]> {
  if (!files.length || files.length > MAX_DOCUMENTS) throw new ValidationError(`Upload between 1 and ${MAX_DOCUMENTS} documents.`);
  const total = files.reduce((value, file) => value + file.size, 0);
  if (total > MAX_TOTAL_BYTES) throw new ValidationError("The combined document size exceeds 10 MB.");
  const documents = await Promise.all(files.map((file, index) => extractDocument(file, index + 1)));
  if (documents.reduce((value, item) => value + item.character_count, 0) > MAX_EXTRACTED_CHARACTERS) {
    throw new ValidationError("The documents contain too much text for one analysis. Split them into smaller files.");
  }
  return documents;
}

export function documentAnalysisText(document: ExtractedDocument): string {
  return [
    `DOCUMENT ${document.id}: ${document.name}`,
    ...document.sections.flatMap((section) => [
      `[SOURCE ${section.reference}${section.heading ? ` | ${section.heading}` : ""}]`,
      section.text,
    ]),
  ].join("\n");
}

async function extractDocument(file: File, index: number): Promise<ExtractedDocument> {
  const name = basename(file.name || "document").trim();
  const extension = extname(name).toLocaleLowerCase("en-US");
  if (![".doc", ".docx", ".txt"].includes(extension)) {
    throw new ValidationError("Unsupported document type. Upload Word (.doc or .docx) or plain text (.txt).");
  }
  if (!file.size) throw new ValidationError(`${name} is empty.`);
  if (file.size > MAX_DOCUMENT_BYTES) throw new ValidationError(`${name} exceeds the 5 MB file limit.`);
  const bytes = Buffer.from(await file.arrayBuffer());
  const id = `document-${index}`;
  const sections = extension === ".docx"
    ? await docxSections(id, name, bytes)
    : extension === ".doc"
      ? await docSections(id, name, bytes)
      : textSections(id, name, decodeText(name, bytes));
  const characterCount = sections.reduce((value, item) => value + item.text.length, 0);
  if (characterCount < 20) throw new ValidationError(`${name} does not contain enough readable text to analyze.`);
  if (characterCount > MAX_EXTRACTED_CHARACTERS) throw new ValidationError(`${name} contains too much text for one analysis.`);
  return {
    id,
    name,
    format: extension.slice(1) as ExtractedDocument["format"],
    size_bytes: bytes.length,
    sha256: createHash("sha256").update(bytes).digest("hex"),
    character_count: characterCount,
    section_count: sections.length,
    sections,
  };
}

async function docxSections(id: string, name: string, bytes: Buffer): Promise<DocumentSection[]> {
  if (bytes[0] !== 0x50 || bytes[1] !== 0x4b) throw new ValidationError(`${name} is not a valid DOCX document.`);
  try {
    const archive = await JSZip.loadAsync(bytes, { checkCRC32: true });
    const members = Object.values(archive.files);
    if (members.length > MAX_DOCX_MEMBERS) throw new ValidationError(`${name} contains too many embedded files.`);
    let uncompressed = 0;
    for (const member of members) {
      if (member.dir) continue;
      const size = (member as unknown as { _data?: { uncompressedSize?: number } })._data?.uncompressedSize;
      if (!Number.isSafeInteger(size) || (size ?? 0) < 0) {
        throw new ValidationError(`${name} contains an unreadable ZIP member.`);
      }
      uncompressed += size ?? 0;
      if (uncompressed > MAX_DOCX_UNCOMPRESSED_BYTES) throw new ValidationError(`${name} expands beyond the safe DOCX limit.`);
    }
    const result = await mammoth.extractRawText({ buffer: bytes });
    return textSections(id, name, result.value);
  } catch (error) {
    if (error instanceof ValidationError) throw error;
    throw new ValidationError(`${name} could not be read as DOCX.`);
  }
}

async function docSections(id: string, name: string, bytes: Buffer): Promise<DocumentSection[]> {
  if (!bytes.subarray(0, 8).equals(Buffer.from("d0cf11e0a1b11ae1", "hex"))) {
    throw new ValidationError(`${name} is not a valid legacy Word document.`);
  }
  const directory = await mkdtemp(join(tmpdir(), "tasklattice-doc-"));
  const source = join(directory, "source.doc");
  try {
    await writeFile(source, bytes, { mode: 0o600 });
    const result = await execute("antiword", [source], { timeout: 10_000, maxBuffer: MAX_EXTRACTED_CHARACTERS * 4 });
    return textSections(id, name, result.stdout);
  } catch {
    throw new ValidationError("Legacy .doc conversion is unavailable. Save the file as .docx and retry.");
  } finally {
    await rm(directory, { recursive: true, force: true });
  }
}

function textSections(id: string, _name: string, raw: string): DocumentSection[] {
  const lines = raw.split(/\r?\n/);
  const sections: DocumentSection[] = [];
  for (let start = 0; start < lines.length; start += 20) {
    const text = cleanText(lines.slice(start, start + 20).join("\n"));
    if (!text) continue;
    sections.push({ reference: `${id}:lines-${start + 1}-${Math.min(start + 20, lines.length)}`, heading: "", text });
  }
  return sections;
}

function decodeText(name: string, bytes: Buffer): string {
  for (const encoding of ["utf-8", "utf-16le", "gb18030"] as const) {
    try {
      const value = new TextDecoder(encoding, { fatal: true }).decode(bytes);
      if (!value.includes("\0")) return value.replace(/^\uFEFF/, "");
    } catch { /* try the next reviewed encoding */ }
  }
  throw new ValidationError(`${name} is not valid UTF-8, UTF-16, or GB18030 text.`);
}

function cleanText(value: string): string {
  return value.split(/\r?\n/).map((line) => line.replace(/[ \t]+/g, " ").trim()).filter(Boolean).join("\n").trim();
}
